src/ai/ChatHistorySource.js

```diff
@@ -0,0 +1,44 @@
+import { ContextSource } from './ContextSource.js';
+
+/**
+ * @class ChatHistorySource
+ * This stores the history of messages (user messages, assistant responses, and system function call messages)
+ * so that they can be provided as context to the ChatInterface.
+ * @todo: still needs some way to limit the max number of tokens sent as context
+ */
+export class ChatHistorySource extends ContextSource {
+    constructor() {
+        super('ChatHistory');
+
+        this.history = [];
+
+        // TODO: implement something like this.
+        //  But not too short, or a complex sequence of function calls might get cropped out too soon.
+        // this.PAST_MESSAGES_INCLUDED = 20;
```

Comment:
You don't want to do it based on number of messages, but rather total length of past messages, cropping until you get below some size limit, but at least one past message (In case the last message was huge). This ensures price per request stays roughly as expected.

Reply:
Good idea, I'll move forward with that approach